### ISSUE TITLE
Uploading photos via web service

### ISSUE THREAD

**Issue by Blackbird714** [jeffpiazza/derbynet#106]
I have been trying to get the **photostand.sh** and **shoot-then-scan.sh** scripts to work with my hardware and cameras, and have run into a lot of challenges.

The latest obstacle, which, I could've sworn was working earlier in the week, seems to be related to cookies generated by the login web service are stored as expired, causing subsequent calls to fail like:  `<failure code="notauthorized">Not authorized -- please see race coordinator</failure>`

During login, I see this:

```
< HTTP/1.1 200 OK
< Server: nginx/1.14.2
< Date: Fri, 17 Jan 2020 17:36:20 GMT
< Content-Type: text/xml; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
* Added cookie PHPSESSID="glnm46qsf4cvaqb4qfl8336p4q" for domain wp197dn-01, path /, expire 0
< Set-Cookie: PHPSESSID=glnm46qsf4cvaqb4qfl8336p4q; path=/
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
```

Note that crazy expiration date:  Expires: Thu, 19 Nov 1981 08:52:00 GMT.  That 'expire 0' may be relevant.  When the photo.upload web service is invoked using the cookie:

```
< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Server: nginx/1.14.2
Server: nginx/1.14.2
< Date: Fri, 17 Jan 2020 17:46:49 GMT
Date: Fri, 17 Jan 2020 17:46:49 GMT
< Content-Type: text/xml; charset=utf-8
Content-Type: text/xml; charset=utf-8
< Transfer-Encoding: chunked
Transfer-Encoding: chunked
< Connection: keep-alive
Connection: keep-alive
* Added cookie PHPSESSID="skpcocfl43200is197lhj0c9jn" for domain wp197dn-01, path /, expire 0
< Set-Cookie: PHPSESSID=skpcocfl43200is197lhj0c9jn; path=/
Set-Cookie: PHPSESSID=skpcocfl43200is197lhj0c9jn; path=/
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
Pragma: no-cache
```
This call results in code="notauthorized" and tries to set a new expired cookie.

I've been looking at the scripts and so far, not finding a reason for this to not work.  I've distilled the steps to reproduce down to these commands from my the Raspberry Pi:

```
rm cookie.txt; touch cookie.txt
curl -v --cacert /boot/server.pem --location --show-error -b ./cookie.txt -c ./cookie.txt --data "action=login&name=Photo&password=flashbulb" https://server/derbynet/action.php
curl -v --cacert /boot/server.pem --location -b ./cookie.txt -c ./cookie.txt --show-error --dump-header /dev/stdout -F action=photo.upload -F MAX_FILE_SIZE=30000000 -F repo=car -F barcode=PWDid088 -F autocrop=0 -F "photo=@/home/pi/derbynet/var/photos-2020-01-17/Carid088.jpg;type=image/jpeg" https://server/derbynet/action.php
```

Other details:

```
$ uname -a
Linux server 4.19.75-v7l+ #1270 SMP Tue Sep 24 18:51:41 BST 2019 armv7l GNU/Linux
$ cat /etc/os-release
PRETTY_NAME="Raspbian GNU/Linux 10 (buster)"
$ curl --version
curl 7.64.0 (arm-unknown-linux-gnueabihf) libcurl/7.64.0 OpenSSL/1.1.1d zlib/1.2.11 libidn2/2.0.5 libpsl/0.20.2 (+libidn2/2.0.5) libssh2/1.8.0 nghttp2/1.36.0 librtmp/2.3
Release-Date: 2019-02-06
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smb smbs smtp smtps telnet tftp
Features: AsynchDNS IDN IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL

```

I'm probably missing something silly related to changes that I have made... but hoping somebody can point me in the right direction.  Thanks in advance!

**Comment by jeffpiazza** [comment]
Is the clock set correctly on the server and/or client?

**Comment by Blackbird714** [comment]
Both ends are Raspberry Pi's, which I'm relatively new to using.  There is no hardware clock, but the OS is reporting the correct current time on both devices:

```
$ date
Fri 17 Jan 2020 01:15:12 PM EST
```

**Comment by Blackbird714** [comment]
Based on this line of discussion, I went ahead and setup ntp, made sure it was in sync.  Also, reverted to http (instead of https).  Unfortunately seeing the same behavior.

**Comment by jeffpiazza** [comment]
The cookie and Expires: header may well just be a red herring.
What's in your /etc/derbynet.conf and /boot/derbynet.conf, if you have them?

**Comment by Blackbird714** [comment]
Contents of /etc/derbynet.conf:

```
DERBYNET_SERVER="https://server/derbynet"
ADDRESS=server-Photos
PHOTO_CAPTURE=fswebcam
GPHOTO2_OPTS="--camera=canon"
FSWEBCAM_ARGS="--config /boot/fswebcam.conf"
CURL_OPTS="--cacert /boot/server.pem"
PHOTO_DIR=/home/pi/derbynet/var/photos-$(date '+%Y-%m-%d')
PHOTO_USER=Photo
PHOTO_PASSWORD=flashbulb
BARCODE_SCANNER_DEVS="usb-13ba_Barcode_Reader-event-kbd"
```
I've been changing things -- as you can see from some of the above params.  So this may be self inflicted.  My objective is to get these scripts working headless via systemd -- but mostly sent and forget.  If you're interested, I can share once I have all the kinks worked out...

**Comment by Blackbird714** [comment]
The expires=0 setting was bugging me, so I found the setting in PHP that controls that (session.cookie_lifetime).  I can see the affect of that change:

```
* upload completely sent off: 42 out of 42 bytes
< HTTP/1.1 200 OK
< Server: nginx/1.14.2
< Date: Fri, 17 Jan 2020 18:44:19 GMT
< Content-Type: text/xml; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
* Added cookie PHPSESSID="9cdntkqc5v6gpmhd21k79slrr5" for domain wp197dn-01, path /, expire 1579290259
< Set-Cookie: PHPSESSID=9cdntkqc5v6gpmhd21k79slrr5; expires=Fri, 17-Jan-2020 19:44:19 GMT; Max-Age=3600; path=/
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
<
<?xml version="1.0" encoding="UTF-8"?>
<action-response action="login" name="Photo" password="...">
<timecheck>2020-01-17T13:44:19-0500</timecheck>
<success>Photo</success>
</action-response>
```
But no affect on the behavior:
```
<?xml version="1.0" encoding="UTF-8"?>
<action-response action="photo.upload" MAX_FILE_SIZE="30000000" repo="car" barcode="PWDid088" autocrop="0"><failure code="notauthorized">Not authorized -- please see race coordinator</failure></action-response>
```

**Comment by jeffpiazza** [comment]
Maybe try adding something like var_dump($_SESSION); above line 42 (not_authorized_body();) in website/ajax/action.photo.upload.inc.  

$_SESSION['permissions'] & 8192

is presumably producing zero, but I don't know why.

**Comment by Blackbird714** [comment]
Done; here is the output of the session dump:
```
<
<?xml version="1.0" encoding="UTF-8"?>
<action-response action="photo.upload" MAX_FILE_SIZE="30000000" repo="car" barcode="PWDid088" autocrop="0">array(2) {
  ["role"]=>
  string(0) ""
  ["permissions"]=>
  int(1)
}
<failure code="notauthorized">Not authorized -- please see race coordinator</failure></action-response>
```

The cookie generated by the login was overwritten with a new one created by the photo.upload call.

**Comment by jeffpiazza** [comment]
This says that the session being used (which is identified by the cookie) is one for a not-logged-in user: the "role" name is the empty string, and only the VIEW_RACE_RESULTS_PERMISSIONS bit is set.

See if you can capture what gets returned from the login request.

Also, any chance you made changes to website/local/config-roles.inc?

**Comment by Blackbird714** [comment]
I did not make any changes to the **config-roles.inc file**; just verified that the password matches the default:

```
                'Photo' => array('password' => 'flashbulb',
                                 'permissions' => PHOTO_UPLOAD_PERMISSION,
                                 'interactive' => false),
```

I hacked the login ajax file similarly; here is a fresh login and photo.upload:
```
* upload completely sent off: 42 out of 42 bytes
< HTTP/1.1 200 OK
< Server: nginx/1.14.2
< Date: Fri, 17 Jan 2020 20:08:05 GMT
< Content-Type: text/xml; charset=utf-8
< Transfer-Encoding: chunked
< Connection: keep-alive
* Added cookie PHPSESSID="na2mqmb4qis8iqs6fdftkdog5k" for domain wp197dn-01, path /, expire 1579295285
< Set-Cookie: PHPSESSID=na2mqmb4qis8iqs6fdftkdog5k; expires=Fri, 17-Jan-2020 21:08:05 GMT; Max-Age=3600; path=/
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
<
<?xml version="1.0" encoding="UTF-8"?>
<action-response action="login" name="Photo" password="...">
<timecheck>2020-01-17T15:08:05-0500</timecheck>
array(2) {
  ["role"]=>
  string(5) "Photo"
  ["permissions"]=>
  int(8192)
}
<success>Photo</success>
</action-response>

< HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Server: nginx/1.14.2
Server: nginx/1.14.2
< Date: Fri, 17 Jan 2020 20:09:20 GMT
Date: Fri, 17 Jan 2020 20:09:20 GMT
< Content-Type: text/xml; charset=utf-8
Content-Type: text/xml; charset=utf-8
< Transfer-Encoding: chunked
Transfer-Encoding: chunked
< Connection: keep-alive
Connection: keep-alive
* Added cookie PHPSESSID="k0pp0v0js8itnjc6rm73vpv52m" for domain wp197dn-01, path /, expire 1579295360
< Set-Cookie: PHPSESSID=k0pp0v0js8itnjc6rm73vpv52m; expires=Fri, 17-Jan-2020 21:09:20 GMT; Max-Age=3600; path=/
Set-Cookie: PHPSESSID=k0pp0v0js8itnjc6rm73vpv52m; expires=Fri, 17-Jan-2020 21:09:20 GMT; Max-Age=3600; path=/
< Expires: Thu, 19 Nov 1981 08:52:00 GMT
Expires: Thu, 19 Nov 1981 08:52:00 GMT
< Cache-Control: no-store, no-cache, must-revalidate
Cache-Control: no-store, no-cache, must-revalidate
< Pragma: no-cache
Pragma: no-cache

<
<?xml version="1.0" encoding="UTF-8"?>
<action-response action="photo.upload" MAX_FILE_SIZE="30000000" repo="car" barcode="PWDid088" autocrop="0">array(2) {
  ["role"]=>
  string(0) ""
  ["permissions"]=>
  int(1)
}
<failure code="notauthorized">Not authorized -- please see race coordinator</failure></action-response>
```

**Comment by jeffpiazza** [comment]
It looks like you're starting a new PHP session with each call, but you shouldn't be.

https://stackoverflow.com/questions/19692157/session-variables-not-working-php suggests looking at the directory permissions associated with the sessions directory.  The about.php page can tell you what the session.save_handler value is, but I'm not sure what the default is if the value shows "no value".  /var/lib/php/sessions on debian 4.9.

You might try just rebooting the server if you just want to get past this and not worry about how we got here.

**Comment by Blackbird714** [comment]
I've rebooted both client and server multiple times; I even went back to the stock scripts from derbynet-extras.  Same symptom there too.

I checked permissions to the PHP sessions path configured for the server:
```
$ ls -ld /var/lib/php/sessions
drwx-wx-wt 2 root root 4096 Jan 17 19:59 /var/lib/php/sessions
```
Every time somebody/something connects to the derbynet server, I see a session file created and owned by www-data.  So that suggests that there isn't a permissions issue.

I'll keep poking around as time allows, but I am out of ideas for now.  Thanks for the suggestions, I'll post back if I ever find root cause.

**Comment by Blackbird714** [comment]
I went around the world to figure it out, but I finally determined the root cause here.  I was trying to use SSL with a self-signed certificate.  I was feeding curl a pem file to satisfy the SSL verification requirement.  The webservice call was to a hostname resolved by a router (e.g. server123).  Disabling SSL and removing the pem file didn't change the problem behavior.

Long story short, this article finally drove me to the solution:  https://github.com/curl/curl/issues/3676

When I changed the web service call to connect to a) an IP address; or b) a **hostname.domainname** instead of just **hostname**, the cookie was sent over by curl allowing the **photo.upload** call to succeed.

Hopefully my pain can help someone else down the road!
